prefetch.bs

```diff
@@ -0,0 +1,219 @@
+</div>
+
+<div algorithm="find a matching prefetch response">
+    To <dfn export>find a matching prefetch response</dfn> given a [=document=] |document|, [=URL=] |url| and [=referrer policy=] |referrerPolicy|, perform the following steps.
+
+    1. [=Assert=]: |document| is [=Document/fully active=].
+    1. Let |currentTime| be the [=current high resolution time=] for the [=relevant global object=] of |document|.
+    1. [=list/For each=] |record| of |document|'s [=prefetch buffer=]:
+        1. If |record|'s [=prefetch record/URL=] is not equal to |url| or |record|'s [=prefetch record/referrer policy=] is not equal to |referrerPolicy|, then [=iteration/continue=].
+        1. [=list/Remove=] |record| from |document|'s [=prefetch buffer=].
+        1. If |record|'s [=prefetch record/expiry time=] is less than |currentTime|, return null.
+        1. Return |record|'s [=prefetch record/response=].
+    1. Return null.
+    
+    <p class="issue">It might be possible to use cache response headers to determine when a response can be used multiple times, but given the short lifetime of the prefetch buffer it's unclear whether this is worthwhile.</p>
+</div>
+
+<h2 id="prefetch-algorithms">Prefetch algorithms</h2>
+
+These algorithms are based on [=process a navigate fetch=].
+
+<p class="issue">Check whether we need to expressly call Should request/response by blocked by Content Security Policy</p>
+<p class="issue">Check Service Worker integration</p>
+
+<div algorithm="partitioned prefetch">
+    To <dfn export>partitioned prefetch</dfn> given a [=document=] |document|, [=URL=] |url| and [=referrer policy=] |referrerPolicy|, perform the following steps.
+
+    1. [=Assert=]: |url|'s [=url/scheme=] is an [=HTTP(S) scheme=].
+    1. Let |partitionKey| be the result of [=determining the network partition key=] given |document|'s [=relevant settings object=].
+    1. Let |browsingContext| be [=document=]'s [=Document/browsing context=].
+    1. Let |request| be a [=request=] as follows:
+
+        :  [=request/URL=]
+        :: |url|
+        :  [=request/referrer policy=]
+        :: |referrerPolicy|
+        :  [=request/initiator=]
+        :: "`prefetch`"
+        :  [=request/destination=]
+        :: "`document`"
+        :  [=request/mode=]
+        :: "`navigate`"
+        :  [=request/credentials mode=]
+        :: "`include`"
+        :  [=request/redirect mode=]
+        :: "`manual`"
+        :  [=request/use-URL-credentials flag=]
+        :: (set)
+        :  [=request/client=]
+        :: |document|'s [=relevant settings object=]
+        :  [=request/header list=]
+        ::
+            *  `` `Sec-Purpose` ``/`` `Prefetch` ``
+    1. Let |locationURL| be null.
+    1. Let |currentURL| be |request|'s [=request/current URL=].
+    1. While true:
```

Comment:
Oof, duplicating all the navigation-redirect-handling logic here is unfortunate, especially since we keep finding bugs in it and fixing it. (E.g. https://github.com/whatwg/html/commit/431e8584c95d4a5483cb4ed592e65c668d701984)

I'd suggest either trying to create a refactoring so it can be shared (i.e. spec a shared algorithm, then spec the modifications to process a navigate fetch to use your shared algorithm), or at least have a TODO to deduplicate.

Reply:
I can spend some more time trying to. I'm trying to keep it vaguely readable as opposed to "like process a navigate fetch, but with tons of diffs".

I'm worried that if I try to abstract to deduplicate it may harm readability further, but it's worth a shot.